segmentstore/server/src/main/java/io/pravega/segmentstore/server/attributes/SegmentAttributeBTreeIndex.java

```diff
@@ -502,15 +515,63 @@ private long deserializeValue(ByteArraySegment value) {
                         "Attribute Index Segment has not been created yet. Cannot read %d byte(s) from offset (%d).",
                         length, offset)));
             }
-        } else {
-            byte[] buffer = new byte[length];
-            return this.storage.read(handle, offset, buffer, 0, length, timeout)
-                               .thenApplyAsync(bytesRead -> {
-                                   Preconditions.checkArgument(length == bytesRead, "Unexpected number of bytes read.");
-                                   storeInCache(offset, buffer);
-                                   return new ByteArraySegment(buffer);
-                               }, this.executor);
         }
+
+        return readPageFromStorage(handle, offset, length, timeout);
+    }
+
+    /**
+     * Reads a BTreeIndex page from Storage and inserts its into the cache (no cache lookups are performed).
+     * Handles read concurrency on the same page by piggybacking on existing running reads on that page (identified by
+     * page offset). If more than one concurrent request is issued for the same page, only one will be sent to Storage,
+     * and subsequent ones will be attached to the original one.
+     *
+     * @param handle  {@link SegmentHandle} to read from.
+     * @param offset  Page offset.
+     * @param length  Page length.
+     * @param timeout Timeout for the operation.
+     * @return A CompletableFuture that will contain the result.
+     */
+    @VisibleForTesting
+    CompletableFuture<ByteArraySegment> readPageFromStorage(SegmentHandle handle, long offset, int length, Duration timeout) {
+        PendingRead pr;
+        synchronized (this.pendingReads) {
+            pr = this.pendingReads.get(offset);
+            if (pr == null) {
+                // Nobody else waiting for this offset. Register ourselves.
+                pr = new PendingRead(offset, length);
+                pr.completion.whenComplete((r, ex) -> {
+                    // Cleanup.
+                    synchronized (this.pendingReads) {
+                        this.pendingReads.remove(offset);
+                    }
+                });
+                this.pendingReads.put(offset, pr);
+            } else if (pr.length < length) {
+                // Somehow the previous request wanted to read less than us. This shouldn't be the case, yet it is
+                // a situation we should handle. In his case, we will not be recording the PendingRead.
+                pr = new PendingRead(offset, length);
+            } else {
+                // Piggyback on the existing read.
+                return pr.completion;
```

Comment:
What happens when page is evicted ?
I think here when page is evicted from cache we need to remove that offset from pendingReads , otherwise next time we need to reload the page again we'll have a stale completed future from the last time and we won't really read anything - leaving cache with missing data.

Reply:
Oops my bad. 
Never mind - I now see it is removed on line 546. Thanks